src/views/Timesheets/index.js

```diff
@@ -0,0 +1,539 @@
+          nextDate = 1;
+          monthToReturn = (date.getMonth() + 1) % 12;
+          yearToReturn = date.getFullYear();
+        } else {
+          nextDate = date.getDate() + 1;
+          monthToReturn = date.getMonth();
+          yearToReturn = date.getFullYear();
+        }
+      }
+
+      return {
+        date: nextDate,
+        month: monthToReturn,
+        year: yearToReturn,
+      };
+    };
+
+    const disableDisallowedDays = date => {
+      let dayIn = selectedDateIn;
+      let nextDate = getNextDate(dayIn);
+      let shouldDisableDate = !(
+        (date.getDate() === dayIn.getDate() &&
+          date.getMonth() === dayIn.getMonth() &&
+          date.getYear() === dayIn.getYear()) ||
+        (date.getDate() === nextDate.date &&
+          date.getMonth() === nextDate.month &&
+          date.getFullYear() === nextDate.year)
+      );
+      return shouldDisableDate;
+    };
+
+    const handleCloseSnackbar = (event, reason) => {
+      if (reason === 'clickaway') {
+        return;
+      }
+
+      setSnackbarOpen(false);
+    };
+
+    const jobDropdown = (
+      <FormControl
+        disabled={userJobs === null || userJobs.length === 0}
+        style={{
+          width: 252,
+        }}
+      >
+        <InputLabel>Jobs</InputLabel>
+        <Select
+          value={selectedJob}
+          onChange={e => {
+            setSelectedJob(e.target.value);
+          }}
+          input={<Input id="job" />}
+        >
+          <MenuItem label="None" value="">
+            <em>None</em>
+          </MenuItem>
+          {jobsMenuItems}
+        </Select>
+      </FormControl>
+    );
+
+    let errorText;
+    if (enteredFutureTime) {
+      errorText = (
+        <Typography variant="overline" color="error">
+          A shift cannot begin or end in the future.
+      </Typography>
+      );
+    } else if (timeOutIsBeforeTimeIn) {
+      errorText = (
+        <Typography variant="overline" color="error">
+          A shift cannot end before it starts.
+      </Typography>
+      );
+    } else if (shiftTooLong) {
+      errorText = (
+        <Typography variant="overline" color="error">
+          A shift cannot be longer than 20 hours.
+      </Typography>
+      );
+    } else if (shiftTooShort) {
+      errorText = (
+        <Typography variant="overline" color="error">
+          A shift cannot be shorter than 15 minutes.
+      </Typography>
+      );
+    } else if (isOverlappingShift) {
+      errorText = (
+        <Typography variant="overline" color="error">
+          The entered shift conflicts with a previous shift.
+      </Typography>
+      );
+    }
+    else {
+      errorText = <></>;
+    }
+
+    const onDatetimeSelectorClose = () => {
+      setIsOverlappingShift(false);
+      if (selectedDateIn !== null && selectedDateOut !== null) {
+        getActiveJobsForUser();
+        checkForFutureDate();
+      }
+    }
+
+    const handleShiftNotesChanged = event => {
```

Comment:
Is there any chance of the shift notes input being malicious?

Reply:
I'm fairly certain there's no chance, but it's worth doing a bit of research on it. Here's what happens in the back end:

```c#
IEnumerable<StudentTimesheetsViewModel> result = null;

var id_num = new SqlParameter("@ID_NUM", studentID);
var eml = new SqlParameter("@eml", jobID);
var shiftStartDateTime = new SqlParameter("@shift_start_datetime", shiftStart);
var shiftEndDateTime = new SqlParameter("@shift_end_datetime", shiftEnd);
var hours_worked = new SqlParameter("@hours_worked", hoursWorked);
var notes = new SqlParameter("@shift_notes", shiftNotes);
var changedBy = new SqlParameter("@last_changed_by", lastChangedBy);

try
           {
               result = RawSqlQuery<StudentTimesheetsViewModel>.StudentTimesheetQuery("student_timesheets_insert_shift @ID_NUM, @eml, @shift_start_datetime, @shift_end_datetime, @hours_worked, @shift_notes, @last_changed_by", id_num, eml, shiftStartDateTime, shiftEndDateTime, hours_worked, notes, changedBy);
           }
```

My understanding is that because shiftNotes is passed in as a parameter instead of explicitly being part of the query, it is safe from malicious input. But again, we really should research this first. Thanks for bringing it up!

Reply:
After looking this up, I found out that `SqlParameter` does, in fact, prevent SQL injection. It doesn't prevent xss, which is a vulnerability that I'm not sure we prevent anywhere on the site. That's important, and it should probably be an item on the list for the summer practicum. if it happens this summer.

Reply:
Good!